src/idl/src/annotation.c

```diff
@@ -510,6 +510,299 @@ annotate_must_understand(
   return IDL_RETCODE_OK;
 }
 
+static idl_floatval_t idl_arithmetic_to_double(const idl_literal_t *lit) {
+  if (idl_type(lit) & IDL_INTEGER_TYPE) {
+    idl_intval_t val = idl_intval(lit);
+    if (IDL_UNSIGNED & val.type)
+      return (idl_floatval_t)val.value.ullng;
+    else
+      return (idl_floatval_t)val.value.llng;
+  } else if (idl_type(lit) & IDL_FLOATING_PT_TYPE) {
+    return idl_floatval(lit);
+  } else {
+    assert(0);
+  }
+  return 0;
+}
+
+static idl_retcode_t check_and_attach_minmax(
+  idl_pstate_t *pstate,
+  idl_annotation_appl_t *parent,
+  idl_annotation_appl_param_t *param,
+  idl_node_t *node,
+  bool to_max)
+{
+  idl_type_t param_type, field_type;
+  const idl_const_expr_t *const_expr;
+  const idl_annotation_appl_t *existing_for_field;
+
+  assert(param);
+  const_expr = param->const_expr;
+
+  assert(idl_is_literal(const_expr));
+  param_type = idl_type(const_expr);
+
+  if (idl_is_member(node)) {
+    if (to_max)
+      existing_for_field = ((idl_member_t*)node)->max.annotation;
+    else
+      existing_for_field = ((idl_member_t*)node)->min.annotation;
+    field_type = idl_type(((idl_member_t*)node)->type_spec);
+  } else if (idl_is_case(node)) {
+    if (to_max)
+      existing_for_field = ((idl_case_t*)node)->max.annotation;
+    else
+      existing_for_field = ((idl_case_t*)node)->min.annotation;
+    field_type = idl_type(((idl_case_t*)node)->type_spec);
+  } else {
+    idl_error(pstate, idl_location(param),
+      "@max/@min/@range values can only be assigned to members and cases");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+
+  if (existing_for_field) {
+    idl_error(pstate, idl_location(param),
+      "attempting to overwrite existing limit value");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+
+  if (!(field_type & IDL_INTEGER_TYPE) && !(field_type & IDL_FLOATING_PT_TYPE) && field_type != IDL_OCTET) {
+    idl_error(pstate, idl_location(param),
+      "@max/@min/@range field type mismatch: attempt to set limits on non-arithmetic field");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  } else if (param_type & IDL_FLOATING_PT_TYPE && field_type & IDL_INTEGER_TYPE) {
+    idl_error(pstate, idl_location(param),
+      "@max/@min/@range field type mismatch: attempt to set floating point limit on integer field");
+    return IDL_RETCODE_SEMANTIC_ERROR;
+  }
+
+  const idl_literal_t *lit = (const idl_literal_t *)const_expr;
+  if (param_type == IDL_OCTET || (param_type & IDL_INTEGER_TYPE)) {
+    //check whether range of value falls inside type
+    idl_intval_t intval = idl_intval(lit);
+    bool is_unsigned = intval.type & IDL_UNSIGNED;
+    uint64_t uval = intval.value.ullng;
+    int64_t sval = intval.value.llng;
+
+    if (!is_unsigned && sval < 0 && field_type & IDL_UNSIGNED && (field_type & IDL_FLOATING_PT_TYPE) == 0) {
```

Comment:
... and while we're at it, for consistency with the other places (in this PR anyway 🙂) it seems more logical to use a logical-not rather than equals-to-zero

```suggestion
    if (!is_unsigned && sval < 0 && (field_type & IDL_UNSIGNED) && !(field_type & IDL_FLOATING_PT_TYPE)) {
```

(and yes, this is nitpicking; I wouldn't have done it if this were all, but the non-parenthesised `DDS_XTypes_TRY_CONSTRUCT_TRIM` means I have to do another round anyway)

Reply:
I will fix this.